File: 494.目标和/findTargetSumWays.go
package problem0494

// 01背包
// 将nums划分成2个子集A，B，分别代表分配 + 的数和分配 - 的数，存在如下关系：
// sum(A) - sum(B) = S
// ===> sum(A) = S + sum(B)
// ===> sum(A) * 2 = S + sum(B) + sum(A)
// ===> sum(A) = (S + sum(nums)) / 2
// 原问题转换为：nums中存在几个子集A，使得A中元素和为(S + sum(nums)) / 2
// 按照416解题
// dp[0][...] = 0 ==> 没有物品的话，根本没办法装背包
// dp[...][0] = 1 ==> 没有容量的话，什么都不装就是一种装法
func findTargetSumWays(nums []int, S int) int {
	sum := 0
	for _, v := range nums {
		sum += v
	}
	if S > sum || (S+sum)%2 == 1 {
		return 0
	}
	target := (S + sum) / 2
	dp := [2001]int{1}
	for i := 1; i <= len(nums); i++ {
		for j := target; j >= nums[i-1]; j-- {
			dp[j] += dp[j-nums[i-1]]
		}
	}
	return dp[target]
}

File: 494.目标和/findTargetSumWays.js
// 由于 (sum - 作为正数的数字和) - 作为正数的数字和 = target
// 所以目标和为 target 转换为目标和为 (sum - target) / 2
// 每个元素的状态也从"添加加号"、"添加减号" 变成了 "选"、"不选"
// 问题转换为01背包问题
function findTargetSumWays(nums, target) {
  // 排除基本情况
  const sum = nums.reduce((acc, i) => acc += i, 0)
  if (target > sum) {
    return 0
  }
  const realTarget = (sum - target) / 2
  if (realTarget % 1 !== 0) {
    return 0
  }
  // 直接使用优化空间的dp求解
  const dp = new Array(realTarget + 1).fill(0)
  dp[0] = 1
  for (let i = 1; i <= nums.length; i++) {
    for (let j = realTarget; j >= nums[i - 1]; j--) {
      dp[j] += dp[j - nums[i - 1]]
    }
  }
  return dp[realTarget]
}
